### PR TITLE
Autocomplete Component: add e2e tests (take two)

### PR DESCRIPTION
**Pull request by chad1008**
## What?
Add some basic e2e testing for the Autocomplete component, as well as some tests specific to past regressions

## Why?
The component currently lacks detailed testing, outside of some user mention tests. Increased testing will allow for safer/more confident updates to the component in the future.

## How?
This is an alternate approach based on the initial work and feedback in #42674

There are two specific regressions [introduced here](https://github.com/WordPress/gutenberg/pull/41382) (already reverted) that I want to include tests to avoid. See issues below for more details on those regressions:
- https://github.com/WordPress/gutenberg/issues/41709
- https://github.com/WordPress/gutenberg/issues/41724

These regression appear to effect only certain completers, like user mentions. I _think_ the issue is actually any autocompleter that relies on fetched data to populate its options. I haven't fully confirmed that theory, but the important part is that these regressions don't impact a completer whose options come from a simple locally declared object.

With that in mind I wanted these tests to confirm if a failure was impacting simple completers, more complex completers (like user mentions), or both. To accomplish this I'm running the tests through a `describe.each` to account for all cases. Because the data set (the completer options) is different for the two paths, most of the tests conditionally define their test data within the test itself. This way the testing logic is the same but the data is tailored to each completer. It should also make it relatively easy to test additional completers in the future if needed.

I also had to add a bit of logic for a single test in the `.each` to be skipped for user mentions. I wrapped this logic in a helper function so it can also be easily reused in the future if necessary.

Finally, in addition to creating a custom completer, the testing plugin also creates a clone of the existing user mention completer. This is used in one test to confirm that two complex (i.e. data-fetching) completers used in the same block won't cause the browser to crash (one of the regressions I'm trying to guard against). Unfortunately, while this cloning process works and accurately triggers the regression, the infinite loop at the heart of that regression _also_ crashes our tests 🤦. I haven't found a way around that, so I've left the test skipped for now. We can remove it (I know the linter disapproves of skipped tests) if that's preferable, but for now I opted to keep it in case we do find a way to utilize it while still getting test results in the event that this particular regression is ever reintroduced.

## Testing Instructions
1. Run `npm run test:e2e  -- packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js`
2. All tests should pass
3. Apply the changes from #41382 to re-introduce the regressions. Rebuild Gutenberg.
4. Re-run the tests. You should see one failure:
- User Mentions: should allow newlines after multiple completions

Optional: Finally, if you remove the `.skip` from the `Autocomplete: should insert elements from multiple completers in a single block` test and re-run the suite, you should find that it simply hangs, not returning any data. I'm calling this an optional step because we're really just testing that the test can trigger the regression by breaking your browser, which isn't actually all that helpful 😉

### CODE REVIEW

**Comment by ciampo** [comment]
>  Working on that part now, and then the whole thing will get a polish pass before this PR is ready for review.

Just ping us when it's time!

**Comment by chad1008** [comment]
Marking this as ready for review (cc @ciampo @mirka).

I wasn't ultimately able to find a way around the crash caused by the infinite loop in one of the regressions I wanted to test against. The test in question is skipped, but still included just in case we're ever able to make use of it. Happy to remove it if need be!

File: packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js
Comment (ciampo) [inline packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js:226]: Wouldn't `Buddy Elf` be the first one, in that case?

Comment (ciampo) [inline packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: If pressing arrow down 6 times should select the 7th item in the list and the names are sorted by firt + last name, wouldn't `Katniss` be the target, instead of Jean-Luc ?

File: packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js
Comment (ciampo) [inline packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: We can probably do without this function / abstraction — it used only once and its benefits are not worth disabling `eslint` in a few places.

File: packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js
Comment (ciampo) [inline packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: Without the changes from https://github.com/WordPress/gutenberg/pull/41382, it looks like this test is passing:

Screenshot 2022-09-01 at 10 36 29

It would be a pity not being able to enable this test. Have you tried detecting errors with `page.on('error' ...)` or `page.on('pageerror' ...)` ? Or surrounding some code in this test with a `try/catch` ? This [link](https://github.com/puppeteer/puppeteer/issues/3709) may be of help

**Review by ciampo — COMMENTED** [review]
Props for writing clear tests, well commented, and with semantic assertions!

I confirm that some of the tests fail when adding changes from https://github.com/WordPress/gutenberg/pull/41382, which is great and exactly what this PR is for ✅ 

Apart from inline comments, here's a couple more observations — although we're quite close to being able to merge this one!

### Local failures

For some reason, when I ran `npm run test:e2e -- packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js` on my local machine, tests all timed out.

If then added the `--puppeteer-interactive` flag, and they all passed 🤷 

From that point onwards, running the standard command (i.e. without `--puppeteer-interactive`) also worked.

### E2E vs unit/integration

It still bugs me that we're not able to write these tests as Testing-Library unit tests, but [we've already been through that conversation](https://github.com/WordPress/gutenberg/pull/42674#issuecomment-1197291876).

This component is tightly coupled with the editor as, as @mirka suggests, we should look into potentially moving it to another package (e.g `@wordpress/block-editor`), or at least improving its documentation.

File: packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js
Comment (chad1008) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: Seven users were listed, but they were all of them deceived.

For in secret, another user was made... **one user to rule them all.**

Sorry... I couldn't resist. The test site already has an `admin` user, which is what gets the focus when `@` is typed with no additional text. From there, the down arrows work through our list, landing on the good Captain after the sixth press.

That's pretty counter-intuitive though... I actually had to test it to figure out why the math was off after I read your comment. I'll add a note in the code: [f5b50ac](https://github.com/WordPress/gutenberg/pull/42905/commits/f5b50acf89ccbe8d92329881b0cdd9e0d96ddbc4)

File: packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js
Comment (chad1008) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: Yeah, you're right. When I added it I think I'd planned to use it more. Not worth keeping for a single case. Removed in [90f85c5](https://github.com/WordPress/gutenberg/pull/42905/commits/90f85c5e08dcd321bbfd08e1905f0ed7074c8167)!

File: packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js
Comment (chad1008) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: Agreed, I'd love to be able to test this! So far, I haven't had any luck with `page.on('error')`, `page.on('pageerror')`, `page.on('crash')`, `try/catch` or (after your link above) a `process.on('unhandledRejection',...)` 😢 

@WunderBart I know you're good with e2e puzzles, does any magic for this kind of stuck test come to mind?

File: packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js
Comment (ciampo) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: haha thankj you for the context ! I definitely ignored the existence of `admin` 🤦  All good!

File: packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js
Comment (WunderBart) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: Sorry for the delayed response! I'll look into it now. One question for now, though - any reason we're using Puppeteer for this one? I'm asking because we've been migrating all the E2Es to Playwright, so it would make more sense to me to use Playwright for a new test. IIRC, this is where the migrated tests currently are: [/test/e2e/specs/editor/various](https://github.com/WordPress/gutenberg/tree/1c3bc0ac022310bd0013dd572ba0368c80374806/test/e2e/specs/editor/various)

Comment (WunderBart) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: I don't think we'll able to handle this one, unfortunately. If the browser freezes (due to an endless loop in this case), Puppeteer freezes (totally) as well. I'd argue though that it's better to have a test that freezes the whole E2E job than to introduce that regression again. It just needs appropriate description, but ultimately the test works, right? 😅

/cc @kevin940726 for his opinion on this one. 🙏

Comment (ciampo) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: > I'd argue though that it's better to have a test that freezes the whole E2E job than to introduce that regression again.

Yes, from what I understand the test would pass in normal conditions. It would only enter in an infinite loop when applying the breaking changes from https://github.com/WordPress/gutenberg/pull/41382

Comment (chad1008) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: > One question for now, though - any reason we're using Puppeteer for this one? I'm asking because we've been migrating all the E2Es to Playwright, so it would make more sense to me to use Playwright for a new test. IIRC, this is where the migrated tests currently are: [/test/e2e/specs/editor/various](https://github.com/WordPress/gutenberg/tree/1c3bc0ac022310bd0013dd572ba0368c80374806/test/e2e/specs/editor/various)

Good catch, no specific reason - I wasn't thinking about Playwright at the time and just created the new tests in the same directory as the old user mentions tests.

I've just been looking into moving the new tests in this PR over to Playwright, and hit a bit of a snag. 

Unless I'm misreading things, it looks like some of the utils these new tests rely on haven't yet been migrated over to their Playwright counterparts.  Creating these tests there would mean first migrating (or rewriting tests to avoid) the following utils:
- `activatePlugin`
- `deactivatePlugin`
- `createUser`
- `deleteUser`
- `clickBlockAppender`
- `getEditedPostContent`

As each of those would be a separate PR holding this one up, I'm wondering if it doesn't make more sense to merge the new tests as-is, and migrate them later, rather than hold this one up while the above is sorted out.

Curious what you both think @ciampo @WunderBart

Comment (kevin940726) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: - [`activatePlugin`](https://github.com/WordPress/gutenberg/blob/f62f2a8d0d72aa29d1368883bd566b6444940b03/packages/e2e-test-utils-playwright/src/request-utils/plugins.ts#L35-L54)
- [`deactivatePlugin`](https://github.com/WordPress/gutenberg/blob/f62f2a8d0d72aa29d1368883bd566b6444940b03/packages/e2e-test-utils-playwright/src/request-utils/plugins.ts#L56-L75)
- [`createUser`](https://github.com/WordPress/gutenberg/blob/f62f2a8d0d72aa29d1368883bd566b6444940b03/packages/e2e-test-utils-playwright/src/request-utils/users.ts#L47-L83)
- [`deleteUser`](https://github.com/WordPress/gutenberg/blob/f62f2a8d0d72aa29d1368883bd566b6444940b03/packages/e2e-test-utils-playwright/src/request-utils/users.ts#L85-L100)
- [`getEditedPostContent`](https://github.com/WordPress/gutenberg/blob/f62f2a8d0d72aa29d1368883bd566b6444940b03/packages/e2e-test-utils-playwright/src/editor/get-edited-post-content.ts#L6-L18)

are available in the `e2e-test-utils-playwright` package. We found the `clickBlockAppender` util less helpful and can often be done with a simple click on a role selector (`page.click( 'role=button[name="Add default block"i]' )`).

The recommendation is to write new tests with Playwright, but I agree it has a learning curve (and lacks some documentation for `e2e-test-utils-playwright`) so no worries if you want to write it in Puppeteer first.

However, I think it'll be easier to just start with Playwright since it encourages better practices for writing resilient e2e tests. Migrating tests from Puppeteer works too but often requires shifting the mind around to adapt to a different kind of thinking.

Comment (ciampo) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: @chad1008 would you be up for writing these tests in Playwright? It would be a good way to get started with the testing framework anyway

Comment (chad1008) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: Oh, awesome, thank you @kevin940726!

I'd be happy to move these over to Playwright knowing the utils I need are in place!

Comment (chad1008) [reply packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js]: Updated with my first pass at a Playwright refactor. All tests (including the infinite loop regression) are active and passing locally, but this is my first foray into Playwright so I'm sure there are things I can improve.

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (chad1008) [inline test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js:79]: I wasn't able to find a Playwright equivalent of `describe.each()`, and [it doesn't sound like Playwright yet has a builtin for parameterized tests](https://github.com/microsoft/playwright-test/issues/34)... so I've gone with a basic `forEach` instead.

File: packages/e2e-tests/CHANGELOG.md
Comment (ciampo) [inline packages/e2e-tests/CHANGELOG.md]: Should be moved to the `Unreleased` section

**Review by ciampo — APPROVED** [review]
I'm currently unable to run these tests locally, as [I'm getting an error when trying to tun WordPress on Docker](https://github.com/WordPress/gutenberg/issues/44416)

But tests seem to pass in CI, and code changes LGTM.

Probably better if @WunderBart  and/or @kevin940726 , who are more knowledgeable about PlayWright than I am, also take a look at this PR before merging.

I also appreciated the granularity of the comments, it really helped reviewing the refactor to Playwright 🙏

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: Maybe we can run them in parallel to speed it up a little?

```suggestion
		await Promise.all(
			userList.map( ( user ) => (
				requestUtils.createUser( {
					email: `${ user.username }@example.com`,
					...user,
				} )
			) )
		);
```

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (kevin940726) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js:79]: Seems like this is the [recommended approach](https://playwright.dev/docs/test-parameterize) 👍 !

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: It's [recommended](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/README.md#use-accessible-selectors) to use [role selectors](https://playwright.dev/docs/selectors#role-selector) if possible in tests. The same goes for the rest of the file.

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: Nit: we're now recommending `.poll` for promise-assertions.

```suggestion
			await expect.poll( editor.getEditedPostContent ).toBe(
				testData.snapshot
			);
```

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: Do you know if Playwright will terminate this test if it timeouts?

**Review by kevin940726 — COMMENTED** [review]
Nice job! Just a small thing regarding using [role selectors](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/README.md#use-accessible-selectors) for more readable selectors that don't depend on implementation details. Other than that it looks good to me! 👍

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (chad1008) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: Done! [5bbad6e](https://github.com/WordPress/gutenberg/pull/42905/commits/5bbad6ee6682cb8b5e5f1b9fa912fff6cf49c026)

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (chad1008) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: Great idea. Implemented in [3886ba4](https://github.com/WordPress/gutenberg/pull/42905/commits/3886ba462888da6c4f6fafdb479b498a015dfaa3) (the above suggestion still has the original `for` loop, so I couldn't use that one :)

File: packages/e2e-tests/CHANGELOG.md
Comment (chad1008) [reply packages/e2e-tests/CHANGELOG.md]: [Done](https://github.com/WordPress/gutenberg/pull/42905/commits/c0936b2648f612ae90bf51c2bc334d3d1f9a65f9), thanks!

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (chad1008) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: (I also followed up with [00b0116](https://github.com/WordPress/gutenberg/pull/42905/commits/00b0116a5aae290d31a553376b92650557778820) to rename `optionPath` to `optionText` as that makes more sense now)

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (kevin940726) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: Haha oops, I forgot to delete that 😅 .

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (chad1008) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: No problem! [29d0d57](https://github.com/WordPress/gutenberg/pull/42905/commits/29d0d57f4dd217a51ab253f067aca28328fc68ee)

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (chad1008) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: 🥳 It _does!_ That's great news, I'm glad you thought to ask. I'm updating the code comment... now that we're in Playwright, it looks like this test does actually function the way we want it to!

**Comment by chad1008** [comment]
Thanks all! I've made some updates based on the provided feedback. @WunderBart/@kevin940726 let me know if anything else looks out of place!

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (ciampo) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: That's great!

File: test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js
Comment (kevin940726) [inline test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: Could and should we explicitly specify the `name` for the options?

```suggestion
				page.locator( `role=option[name="Darth Vader yourfather"]` )
```

Comment (chad1008) [reply test/e2e/specs/editor/various/autocomplete-and-mentions.spec.js]: Done! 1ead803

**Review by kevin940726 — APPROVED** [review]
LGTM 👍 💯

**Pull request merged**
